### ISSUE TITLE
[Bug] Local runtime: 9B hybrid (qwen35) launched with FFN→CPU override on 8 GB despite fitting — overhead overestimated ~3.7 GiB, hybrid -ot moves every layer's FFN (2.4× slower gen)

### ISSUE THREAD

**Issue by l3ad3r1** [NousResearch/hermes-agent#113329]
### Bug Description

On an 8 GB card, the managed local runtime launches **Ornith-1.5-9B (arch `qwen35`, hybrid: 24 recurrent + 8 full-attention layers)** at the 64K floor with `--override-tensor blk\.\d+\.ffn_.*\.weight=CPU`, cutting generation speed ~2.4×. The model actually fits entirely on the GPU **within Hermes's own usable-VRAM budget**, so the spill decision is wrong, and when a spill *is* decided, the hybrid placement moves far more than needed.

Two separate problems in `hermes_cli/local_runtime/context_policy.py` (identical on `main` as of 4e9d3c713a):

**1. Launch footprint overestimated by ~3.7 GiB → false spill.** `plan_launch()` prices overhead as `RUNTIME_OVERHEAD_BYTES` (1.5 GiB) + `ub_logits_bytes(n_vocab)` = `2048 × n_vocab × 4` → 1.90 GiB for this model's 248,320 vocab. Hermes's decision for this machine:

| term | estimate |
|---|---|
| weights | 4.97 GiB |
| KV @ 64K (q8_0, FA) | 1.16 GiB |
| fixed overhead | 1.50 GiB |
| ub logits (n_ubatch 2048) | 1.90 GiB |
| **need** | **9.52 GiB** |
| usable budget (`probe_budget(planning=True)`, 8 GiB − 2 GiB margin) | 6.00 GiB |
| → `spill_bytes` | 3.53 GiB |

Measured with the exact launch args Hermes generates (`-c 65536 -ctk q8_0 -ctv q8_0 -fa on -b 2048 -ub 2048`) but no `-ot`: **5,960 MiB (5.82 GiB) resident, peaking at +30 MiB while processing a 15,912-token prompt** — so all non-weight costs combined were ~0.85 GiB vs. 4.55 GiB estimated, and the model fits under the 6.00 GiB budget with the full margin intact.

**2. Hybrid spill placement ignores both the layer kind and the spill size.** `spill_overrides()` says "hybrids push recurrent-layer FFNs", but for `profile.recurrent_layer_count > 0` it returns `blk\.\d+\.ffn_.*\.weight=CPU`, which matches the FFN tensors of **every** block, including the 8 full-attention layers. It's also all-or-nothing: the same pattern is emitted whether `spill_bytes` is 100 MB or 5 GB. (Pure dense models get `[]` and MoE the `_exps` pattern — this only affects hybrids.)

### Steps to Reproduce

1. Windows 11, RTX 3070 Laptop 8 GB, `local_runtime.enabled: true`, runtime `b10964` CUDA.
2. Stage a `qwen35` hybrid ~9B Q4 GGUF (here `Ornith-1.5-9B-uncensored.Q4_K_S.gguf`, 262K native ctx).
3. Start Hermes Desktop; inspect `runtimes/llamacpp/presets.ini` → section has `override-tensor = blk\.\d+\.ffn_.*\.weight=CPU`; `# hermes-decision` has `"spilled": true`.

Estimate breakdown reproduced with:

```python
from hermes_cli.local_runtime import presets
from hermes_cli.local_runtime.hardware import probe_budget
from hermes_cli.local_runtime.context_policy import plan_launch, spill_overrides
prof = presets.profile_from_gguf(presets.read_gguf_header(path))
plan = plan_launch(prof, probe_budget(planning=True))
# moe=False, recurrent_layer_count=24, weights 4.97 GiB, n_vocab 248320
# WindowDecision(window=65536, spill_bytes=3784574976, ...), overhead 3.39 GiB
# spill_overrides(prof) == ['-ot', 'blk\\.\\d+\\.ffn_.*\\.weight=CPU']
```

### Expected Behavior

Model launches fully on GPU (no `-ot`), since measured resident + peak fits the usable budget. If a genuine spill is needed for a hybrid, offload only what the spill requires (and only recurrent-layer FFNs, per the docstring).

### Actual Behavior

A/B on a standalone `llama-server` with the generated preset args, same prompts, temperature 0, nothing else on the GPU:

| config | VRAM (model) | gen, 256-tok answer | 15,912-tok prompt: prefill | 15,912-tok prompt: gen |
|---|---|---|---|---|
| Hermes preset (`-ot` FFN→CPU) | 4,303 MiB | **8.1 tok/s** | 1,421.9 tok/s | **8.6 tok/s** |
| same, no `-ot` | 5,960 MiB | **20.8 tok/s** | 1,607.9 tok/s | **19.2 tok/s** |

The override saves 1.66 GiB of VRAM that wasn't needed and costs 2.2–2.6× generation speed. Because `presets.ini` is regenerated on every cold boot, hand-editing it doesn't stick (same pain point as #106895 §2).

### Affected Component

Local runtime (managed llama.cpp) — `context_policy.py` (`plan_launch`, `ub_logits_bytes`, `spill_overrides`); seen via Hermes Desktop.

### Operating System

Windows 11 Pro (26200), RTX 3070 Laptop 8 GB (WDDM), Ryzen 7 5800H, 16 GB RAM.

### Python Version

3.11.16 (bundled runtime)

### Hermes Version

0.21.3 (3c3ab69abb); llama.cpp runtime b10964 CUDA

### Root Cause Analysis

- `ub_logits_bytes` prices a full `n_ubatch × n_vocab` float32 logits buffer on the GPU. Its docstring says it was calibrated on Qwen3.8 Q4; on this 248K-vocab model the measured total of KV + compute buffers + logits was ~0.85 GiB, so the 1.90 GiB logits term alone exceeds reality, and it stacks on the 1.5 GiB fixed constant. The overestimate grows with vocab size, so large-vocab models on small cards are the ones pushed into false spills.
- `spill_overrides` for hybrids uses a block-agnostic regex and ignores `spill_bytes`.

### Proposed Fix

Suggestions (no patch, since the calibration data isn't mine to change blind):

1. Recalibrate the lean-posture overhead (`RUNTIME_OVERHEAD_BYTES` + `ub_logits_bytes` for `mtp_capable=False`) against large-vocab models, or measure-and-cache the real resident size after the first successful load and use it for subsequent preset generation.
2. In `spill_overrides`, build the hybrid pattern from the recurrent layer indices only (e.g. `blk\.(0|1|2|4|…)\.ffn_.*\.weight=CPU`), as the docstring describes.
3. Size the placement to `spill_bytes`: offload only enough FFN blocks to cover the spill, rather than all of them for any non-zero spill.
4. Related to #106895: a persistent per-model override (e.g. "never spill" / `extra_args`) that survives preset regeneration would let users opt out when they've verified a fit.

Current local workaround: an external watcher that strips the `override-tensor` line from `presets.ini` and restarts the idle router, which Hermes's supervisor respawns from the edited file.

**Comment by KeyArgo** [comment]
Checked both halves against current `main` (`4e9d3c713a3`) rather than guessing, since part 2 is decidable from source and part 1 is not. Findings, for whoever picks this up:

**Part 2 confirmed at source, but the fix is a policy call, not a one-liner.** `spill_overrides()` (`hermes_cli/local_runtime/context_policy.py:223-224`) emits `blk\.\d+\.ffn_.*\.weight=CPU` for any profile with `recurrent_layer_count > 0`, which matches the FFN tensors of every block — the 8 full-attention blocks included — while the docstring says "recurrent-layer FFNs". Two things worth settling before narrowing it:

- The rationale in the same docstring is that recurrent (`n_head_kv == 0`) layers carry no KV worth protecting — but a full-attention layer's FFN weights are equally offloadable; its KV stays resident either way. So the broad pattern is not obviously wrong, and the docstring may be the imprecise half.
- Index-targeted patterns *are* derivable for the launch path: `gguf.py::head_counts_kv()` returns a per-layer list (`full_attention_interval` zeroes the non-attention blocks, `qwen35` included), and `profile_from_gguf()` builds `ModelProfile.layers` in that same order, so index `i` ↔ `blk.i` for GGUF-derived profiles — which is what `presets.generate_presets()` uses. One caveat for any index-based approach: catalog-derived profiles (`catalog.py:112`) concatenate kinds as FULL, SWA, RECURRENT, so their indices are *not* block indices.
- Adequacy hazard: the breadth of the current pattern is what frees the bytes that make the spilled launch fit. The A/B table shows the override saving ~1.66 GiB; narrowing it to recurrent-layer FFNs alone offloads roughly 24/32 of the FFN weights, so a narrower placement needs to be sized against `spill_bytes` (your "offload only what the spill requires") — otherwise it trades a 2.2-2.6x speed loss for an OOM.

**Part 1 cannot be settled from source, and changing it without hardware would be unmeasured.** It is a calibration question: the 1.90 GiB term is `ub_logits_bytes()` = `n_ubatch x n_vocab x 4` (2.03 GB at vocab 248,320), and the only way to decide whether that is an overestimate or your build simply allocates the buffer lazily is a re-measurement on the same 8 GB card with the same build. Narrowing `RUNTIME_OVERHEAD_BYTES` or the logits term from one machine's numbers would move the false-spill boundary for every other card.

Reporting the measured A/B separately from the estimator table — and pinning the exact generated args — is what made both halves checkable. Nice report.

**Comment by socksninja** [comment]
This is suitable for one bounded independent verification because the reported problem crosses a concrete execution/performance boundary: the runtime's launch planner predicts a spill that the actual process does not need, and the resulting tensor override materially changes generation speed.

I can independently verify one hardware/runtime path:
- reproduce the launch planning decision on the specified 8 GB configuration;
- capture Hermes's predicted memory terms and emitted override;
- launch with Hermes-generated args and read back actual resident/peak memory;
- compare generation throughput with and without the emitted CPU-offload override;
- distinguish the planner estimate error from the hybrid placement policy;
- return one reproducible evidence record with exact revision, launch args, measurements, and explicit evidence boundary.

Fixed price: US$99 for one independently verifiable evidence unit. No continuation required.
Payment: https://www.paypal.com/ncp/payment/2URELJ9NE4ZBN

**Comment by kvnloo** [comment]
Split the previous mixed fix after review: draft #120778 now addresses only the mechanically provable hybrid placement bug — `-ot` is generated from the actual recurrent layer indices, so full-attention FFNs are no longer matched when a spill is already required.

I deliberately dropped the proposed non-MTP `ub_logits_bytes` repricing. Current `launch_args()` still emits `-b 2048 -ub 2048`, and the existing test suite pins that posture; the aggregate 5.82 GiB VRAM receipt proves the current estimator is conservative but does not isolate which llama.cpp buffers materialize well enough to justify charging a 512-token term. I think that half should stay open until we have a production allocation receipt / multi-hardware calibration or intentionally change the launch posture alongside the estimator.

So #120778 improves the cost of a genuine hybrid spill without claiming to solve the false-spill calibration half of this issue.
